datajoint/external.py

```diff
@@ -6,6 +6,203 @@
+        raise NotImplementedError(
+            'get_blob method not implemented for', type(self))
+
+    def get_obj(self, hash):
+        ''' returns decoded 'obj' '''
+        raise NotImplementedError(
+            'get_obj method not implemented for', type(self))
+
+    def get(self, hash):
+        ''' returns decoded 'obj' '''
+        return self.get_obj(hash)
+
+
+class RawFileHandler(ExternalFileHandler):
+
+    required = ('location',)
+
+    def __init__(self, store, database):
+        super().__init__(store, database)
+        self.check_required(store, 's3', RawFileHandler.required)
+        self._location = self._spec['location']
+
+    def get_folder(self):
+        return os.path.join(self._location, self._database)
+
+    def put(self, obj):
+        (blob, hash) = self.hash_obj(obj)
+
+        folder = self.get_folder()
+        full_path = os.path.join(folder, hash)
+        if not os.path.isfile(full_path):
+            try:
+                with open(full_path, 'wb') as f:
+                    f.write(blob)
+            except FileNotFoundError:
+                os.makedirs(folder)
+                with open(full_path, 'wb') as f:
+                    f.write(blob)
+
+        return (blob, hash)
+
+    def get_blob(self, hash):
+        full_path = os.path.join(self.get_folder(), hash)
+        try:
+            with open(full_path, 'rb') as f:
+                return f.read()
+        except FileNotFoundError:
+                raise DataJointError('Lost external blob')
+
+    def get_obj(self, hash):
+        return unpack(self.get_blob(hash))
+
+    def get(self, hash):
+        return self.get_obj(hash)
+
+
+class CacheFileHandler(ExternalFileHandler):
+    '''
+    A CacheFileHandler mixin implementation.
+
+    Requires a concrete 'upstream' backend file handling implementation.
+    Should be 1st in inheritance list w/r/t backend implementation - e.g.:
+
+      CachedFoo(CacheFileHandler, OtherFileHandler)
+
+    Will cache objects in config['external-cache']['location'], relying on
+    superclass for coherent get/put operations on the 'reference' blobs.
+
+    Cleanup currently not implemented.
+    '''
+
+    def __init__(self, store, database):
+        super().__init__(store, database)  # initialize non-mixin parameters
+
+        # validate mixin cache parameters
+        if 'external-cache' not in config:
+            raise DataJointError('External Cache is not configured')
+
+        cache_spec = config['external-cache']
+
+        if 'location' not in cache_spec:
+            raise DataJointError(
+                'External Cache configuration missing "location"')
+
+        self._cache_spec = cache_spec
+        self._cache_location = cache_spec['location']
+
+    def get_cache_folder(self):
+        return os.path.join(self._cache_location, self._database)
+
+    def _clean_cache(self):
+        pass  # TODO: implement _clean_cache
+
+    def _put_cache(self, blob, hash):
+
+        self._clean_cache()
```

Comment:
why clean cache at every put?

Reply:
because the logic to clean the cache will be complex and function-worthy -
the thinking was clean cache itself can determine if it needs to actually clean anything
(yes this is a frame - but the alternative is having multiple functions tracking cache cleaning state)